[otlp] Update README for UseOtlpExporter & new experimental flag (#5452)
open-telemetry/opentelemetry-dotnet: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/README.md

  * [LogRecordExportProcessorOptions](#logrecordexportprocessoroptions)
  * [MetricReaderOptions](#metricreaderoptions)
  * [Environment Variables](#environment-variables)
    * [Exporter configuration](#exporter-configuration)
    * [Attribute limits](#attribute-limits)

>     * [Experimental Features](#experimental-features)

Reviewer: ```suggestion
    * [Experimental features](#experimental-features)
```

        .AddAspNetCoreInstrumentation() // Use instrumentation to listen to telemetry
    );

    appBuilder.Services.AddOpenTelemetry()
        .UseOtlpExporter();
    ```

    For details see: [Activity Source](../../docs/trace/customizing-the-sdk/README.md#activity-source).

* The exporter registered by `UseOtlpExporter` will be added as the last
  processor in the pipeline established for logging and tracing.

* `UseOtlpExporter` can only be called once. Subsequent calls will result in a
  `NotSupportedException` being thrown.

* `UseOtlpExporter` cannot be called in addition to signal-specific
  `AddOtlpExporter` methods. If `UseOtlpExporter` is called signal-specific
  `AddOtlpExporter` calls will result in a `NotSupportedException` being thrown.

### Configuring signals when using UseOtlpExporter

`UseOtlpExporter` supports the full set of [environment
variables](#environment-variables) listed below including the signal-specific
overrides and users are encouraged to use this mechanism to configure their
exporters.

A `UseOtlpExporter` overload is provided which may be used to set the protocol
and base URL:

```csharp
appBuilder.Services.AddOpenTelemetry()

>     .UseOtlpExporter(
>         OtlpExportProtocol.HttpProtobuf,
>         new Uri("http://localhost:4318/"));

Reviewer: ```suggestion
    .UseOtlpExporter(OtlpExportProtocol.HttpProtobuf, new Uri("http://localhost:4318/"));
```

    );

    appBuilder.Services.AddOpenTelemetry()
        .UseOtlpExporter();
    ```

    For details see: [Activity Source](../../docs/trace/customizing-the-sdk/README.md#activity-source).

* The exporter registered by `UseOtlpExporter` will be added as the last
  processor in the pipeline established for logging and tracing.

* `UseOtlpExporter` can only be called once. Subsequent calls will result in a
  `NotSupportedException` being thrown.

* `UseOtlpExporter` cannot be called in addition to signal-specific
  `AddOtlpExporter` methods. If `UseOtlpExporter` is called signal-specific
  `AddOtlpExporter` calls will result in a `NotSupportedException` being thrown.

### Configuring signals when using UseOtlpExporter

`UseOtlpExporter` supports the full set of [environment
variables](#environment-variables) listed below including the signal-specific
overrides and users are encouraged to use this mechanism to configure their
exporters.

> [!NOTE]
> In OpenTelemetry .NET environment variable keys are retrieved using
  `IConfiguration` which means they may be set using other mechanisms such as
  defined in appSettings.json or specified on the command-line.

> An `UseOtlpExporter` overload is provided which may be used to set the protocol

Reviewer 2: nit: This should be either: "A `UseOtlpExporter` overload" or "An overload for `UseOtlpExporter`". Right?

"UseOtlpExporter" starts with a vowel but starts with a consonant sound.

> and base endpoint but no other options are currently exposed:

Reviewer 2: I think we should omit the part about only so much being currently exposed as it might come across as we are going to add more. We can always add more overloads in future if need be but it need not be mentioned like this.

```suggestion
and base endpoint:
```

> and base endpoint but no other options are currently exposed:

Reviewer: "base endpoint" is undefined and might be misleading.

Author: So I called it base endpoint because when using `HttpProtobuf` the signal path is added to it. There is this note on there to try and explain that:

![image](https://github.com/open-telemetry/opentelemetry-dotnet/assets/28367120/fd1d18a9-e817-44fd-9237-611817e57ad6)

It is also called `baseEndpoint` on the actual API: https://github.com/open-telemetry/opentelemetry-dotnet/blob/b21da4b074f4fa6a56dfd3c9fa27c58f28867753/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Builder/OpenTelemetryBuilderOtlpExporterExtensions.cs#L57

Questions...

1) Did you notice that "Note" and if not, does that satisfy as an explanation?
2) If unsatisfied, do you think more docs will suffice or should the actual thing on the API be renamed?

Reviewer: I guess most folks call it "base URL/URI", check https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#endpoint-urls-for-otlphttp

![image](https://github.com/open-telemetry/opentelemetry-dotnet/assets/17327289/083979f1-d55a-4951-8b91-c7b22e5aae30)

Normally what I will do is to ask a search engine https://www.google.com/search?q=base+endpoint, and it seems the results are not leading to the right path.

Reviewer: > Questions...
> 
> 1. Did you notice that "Note" and if not, does that satisfy as an explanation?
> 2. If unsatisfied, do you think more docs will suffice or should the actual thing on the API be renamed?

1. I noticed it, and I was confused by "base URL", "base endpoint" and "base address".
2. I don't know, haven't got time to think about this.

Author: I just pushed an update to use "base URL" instead.

```csharp
appBuilder.Services.AddOpenTelemetry()
    .UseOtlpExporter(
        OtlpExportProtocol.HttpProtobuf,

>         new Uri("http://otlp_exporter_base_address/"));

Reviewer: Could this be changed to the OpenTelemetry Collector default endpoint? (which is easier for folks to understand)

>         new Uri("http://otlp_exporter_base_address/"));

Reviewer: What is "base address"? It seems to be undefined.

## Enable OTLP Exporter for all signals

Starting with the `1.8.0` version you can use the cross-cutting
`UseOtlpExporter` extension to simplify registration of the OTLP exporter for
all signals (logs, metrics, and traces).

> [!NOTE]
> The cross cutting extension is currently only available when using the
  `AddOpenTelemetry` extension in the
  [OpenTelemetry.Extensions.Hosting](../OpenTelemetry.Extensions.Hosting/README.md)
  package.

```csharp
appBuilder.Services.AddOpenTelemetry()
    .UseOtlpExporter();
```

The `UseOtlpExporter` has the following behaviors:

* Calling `UseOtlpExporter` automatically enables logging, metrics, and tracing.

>   Calls to `WithLogging`, `WithTracing`, and/or `WithLogging` are NOT required

Reviewer: There are two "WithLogging" here and one needs to be changed to "WithMetrics"?

        );

        appBuilder.Services.AddOpenTelemetry()
            .UseOtlpExporter();
    ```

    For details see: [Meter](../../docs/metrics/customizing-the-sdk/README.md#meter).

    When using `Microsoft.Extensions.Hosting` v8.0.0 or greater (a standard part
    of ASP.NET Core) `Meter`s and `Instrument`s can also be enabled using
    configuration.

    `appSettings.json` metrics configuration example:

    ```json
    {
      "Metrics": {
        "EnabledMetrics": {
          "Microsoft.AspNetCore.*": true,
          "System.*": true,
          "MyCompany.*": true,
        }
      }
    }
    ```

    For details about the built-in metrics exposed by .NET see: [Built-in
    metrics in
    .NET](https://learn.microsoft.com/dotnet/core/diagnostics/built-in-metrics).

  * Tracing

    Trace telemetry is controlled by calling `TracerProviderBuilder.AddSource`
    to listen to
    [ActivitySource](https://learn.microsoft.com/dotnet/api/system.diagnostics.activitysource)s
    emitting traces. Typically instrumentation packages will make this call
    automatically.

    Examples:

    ```csharp

>         appBuilder.Services.AddOpenTelemetry()

Reviewer: Reduce indentation?

```suggestion
    appBuilder.Services.AddOpenTelemetry()
```

## Enable OTLP Exporter for all signals

> A cross-cutting extension was added in v1.8.0 to simplify registration of the

Reviewer 2: Since we haven't released this package yet, I think a better way to put this could be something like this:
```suggestion
Starting `1.8.0` version, you can use the cross-cutting extension to simplify registration of the
```

Author: I pushed a tweak

OTLP exporter for all signals (logs, metrics, and tracing).

> [!NOTE]
> The cross cutting extension is currently only available when using the
  `AddOpenTelemetry` extension in the
  [OpenTelemetry.Extensions.Hosting](../OpenTelemetry.Extensions.Hosting/README.md)
  package.

```csharp
appBuilder.Services.AddOpenTelemetry()
    .UseOtlpExporter();
```

The `UseOtlpExporter` has the following behaviors:

* Calling `UseOtlpExporter` automatically enables logging, metrics, and tracing.

>   Calls to `WithLogging`, `WithTracing`, and/or `WithLogging` are NOT required

Reviewer 2: This sounds a little confusing. We start off saying calls to the "With" extension methods are required and then go on to say however only the telemetry enabled using "With" methods would be exported.

Maybe update this to something like:

> When calling `UseOtlpExporter` method, you don't need to separately call `AddOtlpExporter` method for `WithLogging`, `WithTracing`, and/or `WithMetrics`.

Author: This is kind of tricky. The docs we have that I could find seem to all show "Sdk.Create*" style which really doesn't help for this cross-cutting extension using the hosting style. I decided to show some real examples of how to do it. LMK what you think.

Reviewer 3: `WithLogging` - This is not a known stable way of configuring logging today so I think we should not mention it. 

To keep things simple, what do you guys think about something like this.

```
Using UseOtlpExporter automatically enables the OTLP exporter for logs, metrics, and traces. 
For additional sdk configuration, like adding meters to the MeterProvider, 
incorporating sources to the TracerProvider, or 
configuring resources for all three types of signals, additional setup will required. 
For more information, refer to the details below. (Followed by what we have below)
```

Author: I removed the mention of the "With" calls and simplified it down to just...

> Calling `UseOtlpExporter` automatically enables logging, metrics, and tracing however only telemetry which has been enabled will be exported.

## Environment Variables

The following environment variables can be used to configure the OTLP Exporter
for logs, traces, and metrics.

> [!NOTE]
> In OpenTelemetry .NET environment variable keys are retrieved using
  `IConfiguration` which means they may be set using other mechanisms such as

>   defined in appSettings.json or specified on the command-line.

Reviewer: ```suggestion
  defined in `appSettings.json` or specified on the command-line.
```

* `OTEL_EVENT_ATTRIBUTE_COUNT_LIMIT`
* `OTEL_LINK_ATTRIBUTE_COUNT_LIMIT`

The following environment variables can be used to configure log record limits
used for logging:

* `OTEL_LOGRECORD_ATTRIBUTE_VALUE_LENGTH_LIMIT`
* `OTEL_LOGRECORD_ATTRIBUTE_COUNT_LIMIT`

### Experimental Features

### All signals

* `OTEL_DOTNET_EXPERIMENTAL_OTLP_ENABLE_INMEMORY_RETRY`

>   When set to `true`, it enables in-memory retry for transient errors ecountered

Reviewer 3: ```suggestion
  When set to `true`, it enables in-memory retry for transient errors encountered
```

Author: Fixed, thanks!

* `OTEL_EVENT_ATTRIBUTE_COUNT_LIMIT`
* `OTEL_LINK_ATTRIBUTE_COUNT_LIMIT`

The following environment variables can be used to configure log record limits
used for logging:

* `OTEL_LOGRECORD_ATTRIBUTE_VALUE_LENGTH_LIMIT`
* `OTEL_LOGRECORD_ATTRIBUTE_COUNT_LIMIT`

> ### Experimental features

Reviewer 2: Make this a bigger heading.

```suggestion
## Experimental features
```

Author: Right now it is a sub-section of "Environment Variables" with the toc represented as:

  * [Environment Variables](#environment-variables)
    * [Exporter configuration](#exporter-configuration)
    * [Attribute limits](#attribute-limits)
    * [Experimental features](#experimental-features)

I can pop it out but it will become its own section and then probably needs some text that it is envvar-based. Is that what you want?

Reviewer 2: We have two kinds of experimental features:
- Environment Variables/ Config based
- Public API changes that are enabled by default in pre-release builds

I think we should make "Experimental Features" its own section. The public API ones are not dependent on environment variables. Even though OTLP Exporter only has environment variable based experimental features other components such as SDK have both kinds of experimental features. It would be nice to stay consistent and have all docs contain a dedicated section for experimental features.

Author: Pushed updates: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5452/commits/727012a546d511abd0dc9f5e420e36bb52cdc180

### All signals

* `OTEL_DOTNET_EXPERIMENTAL_OTLP_ENABLE_INMEMORY_RETRY`

  When set to `true`, it enables in-memory retry for transient errors

>   encountered sending telemetry.

Reviewer 2: ```suggestion
  encountered while sending telemetry.
```

```

> [!NOTE]
> `IHttpClientFactory` is NOT currently supported by `OtlpLogExporter`.

## Experimental features

The following features are exposed experimentally in the OTLP Exporter. Features
are exposed experimentally when either the [OpenTelemetry
Specification](https://github.com/open-telemetry/opentelemetry-specification)
has explicitly marked something
[experimental](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/document-status.md)
or when the SIG members are still working through the design for a feature and
want to solicit feedback from the community.

### Environment variables

> [!NOTE]
> In OpenTelemetry .NET environment variable keys are retrieved using
  `IConfiguration` which means they may be set using other mechanisms such as
  defined in appSettings.json or specified on the command-line.

* All signals

>   * `OTEL_DOTNET_EXPERIMENTAL_OTLP_ENABLE_INMEMORY_RETRY`

Reviewer: @vishweshbankwar I have questions regarding the name - does it mean later there will be something like "OTLP_ENABLE_ONDISK_RETRY`? What's the thinking which leads to the current name instead of `OTEL_DOTNET_EXPERIMENTAL_OTLP_ENABLE_RETRY`?

## Experimental features

The following features are exposed experimentally in the OTLP Exporter. Features
are exposed experimentally when either the [OpenTelemetry
Specification](https://github.com/open-telemetry/opentelemetry-specification)
has explicitly marked something
[experimental](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/document-status.md)
or when the SIG members are still working through the design for a feature and
want to solicit feedback from the community.

### Environment variables

> [!NOTE]
> In OpenTelemetry .NET environment variable keys are retrieved using
  `IConfiguration` which means they may be set using other mechanisms such as
  defined in appSettings.json or specified on the command-line.

* All signals

  * `OTEL_DOTNET_EXPERIMENTAL_OTLP_ENABLE_INMEMORY_RETRY`
 
    When set to `true`, it enables in-memory retry for transient errors
    encountered while sending telemetry.

    Added in `1.8.0-beta.1`.

* Logs

  * `OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EVENT_LOG_ATTRIBUTES`

    When set to `true`, it enables export of `LogRecord.EventId.Id` as

>     `logrecord.event.id` and `LogRecord.EventId.Name` as `logrecord.event.name`.

Reviewer 2: Since you have added the starting version for the above feature, maybe add it for this one as well?
